File: py110/small_problems_110_119/advanced/4.py
# https://launchschool.com/exercises/64bea0a1

"""
Q:
1. 

input: two sorted lists
output: list that contains all elements from both list in ascending sorted order
rules:
1. Contain either all integer or string values
2. Cannot sort the result list
3. You must build the list one element at a time
4. should not mutate the input lists
5. Empty list does not matter it will return the other list



2 6 

1 2 5 6 9
while loop


Algo:
1. Create empty list
2. While loop to iterate between both lists. Condition is based on
whether or not there is still elelemtsn in the list(while list1 has elements and
list2 has elements)
    2a. compare the two elements and whichever is less append it first
    2b. append the second element
3. check if list1 still has elements add the rest
4. check if list2 still has elements add the rest
5. return the list


"""

def merge(lst1, lst2):
    result = []
    index1 = 0
    index2 = 0

    while index1 < len(lst1) and index2 < len(lst2):
        num1 = lst1[index1]
        num2 = lst2[index2]

        if num1 <= num2:
            result.append(num1)
            index1 += 1
        else:
            result.append(num2)
            index2 += 1
            
    if index1 < len(lst1):
        result.extend(lst1[index1:])

    if index2 < len(lst2):
        result.extend(lst2[index2:])
    
    return result

# All of these examples should print True
print(merge([1, 5, 9], [2, 6, 8]) == [1, 2, 5, 6, 8, 9])
print(merge([1, 1, 3], [2, 2]) == [1, 1, 2, 2, 3])
print(merge([], [1, 4, 5]) == [1, 4, 5])
print(merge([1, 4, 5], []) == [1, 4, 5])

names1 = ['Alice', 'Kim', 'Pete', 'Sue']
names2 = ['Bonnie', 'Rachel', 'Tyler']
names_expected = ['Alice', 'Bonnie', 'Kim', 'Pete',
                  'Rachel', 'Sue', 'Tyler']
print(merge(names1, names2) == names_expected)

File: ls215/smallProblems/adv/4.js
'use strict';

/*
input: two sorted array
output: return one array that is sorted with both elements from each array
rules:
1. Cant use a sorting mechaninsm.
2. Not mutate the input arrays.

1 5 9.   2 6 8
x x      x x x

i = 2
j = 3

[1, 2, 5, 6, 8, 9]

9 and 8



1 2 5 6 8 9

1 1 3.  1, 2 2

i = 2
j = 1

[1, 1, 1]

3 and 1


Q:


D:
1. A result array

A:
1. Check if either input array is empty. If so return the other array.
2. create result array
3. create ptr that start at 0 to represent each array.
4. While ptr is less than length of input1 and ptr2 is less than length of input2
  5. if elemtn in input1 is less than or equal input 2
    6. add element to result.
    6a. increment ptr 1.
  7. else if input1 is greater to input 2.
    7a. add input2 element to result
    7b. increment ptr2.

8. Check if there are any more elements in input 1
  8a. add all the elements to the result array.

9a. check if htere are any more elements in put 2
  9b. add all the elements to result array

10. return result


*/


function merge(arr1, arr2) {
  if (arr1.length === 0) return arr2;
  if (arr2.length === 0) return arr1;

  let result = [];
  let ptr1 = 0;
  let ptr2 = 0;

  while (ptr1 < arr1.length && ptr2 < arr2.length) {
    if (arr1[ptr1] <= arr2[ptr2]) {
      result.push(arr1[ptr1]);
      ptr1 += 1;
    } else if (arr1[ptr1] > arr2[ptr2]) {
      result.push(arr2[ptr2]);
      ptr2 += 1;
    }
  }

  if (ptr1 < arr1.length) {
    result = result.concat(arr1.slice(ptr1,));
  }

  if (ptr2 < arr2.length) {
    result = result.concat(arr2.slice(ptr2,));
  }

  return result;
}


console.log(merge([1, 5, 9], [2, 6, 8]));      // [1, 2, 5, 6, 8, 9]
console.log(merge([1, 1, 3], [2, 2]));         // [1, 1, 2, 2, 3]
console.log(merge([], [1, 4, 5]));             // [1, 4, 5]
console.log(merge([1, 4, 5], []));             // [1, 4, 5]